Draw support to say the least

kristianserrano opened the pull request:
- Adds `pass`, `deal`, and `createCard` hooks to trigger display.
  - Hooks have logic for whether or not to display based on setting.
  - Adds use of flags on cards to determine whether to render viewer.
- Adds use of card data for values to apply in FancyDisplay template.
- Adds settings for display on pass to hand, pass to pile, and draw.
- Adds `PseudoCard` class for constructing fake card data structure when displaying custom images instead of cards.
- Adds flipping of card in stack when flipped in viewer.
- Removes isFaceDown for replacement by actual `card.showFace` value. Flipping the card in the stack before dealing or passing will set this.
- Updates documentation.

**NOTE: Macros have not yet been updated!**

orcnog committed:
Merge pull request #15 from orcnog/develop

Develop

orcnog committed:
Merge pull request #16 from orcnog/develop

release 0.1.4

orcnog committed:
Merge pull request #17 from orcnog/develop

fixed a typo found by @kristianserrano in PR #14, and version bump. release 0.1.5

orcnog committed:
Fixes issue #18. Adds config setting for borderWidth and adjusts glintEdge if 0

kristianserrano committed:
Refactors module
- Adds pass, deal, and createCard hooks to trigger display.
- Adds use of card data for values in display.
- Adds use of flags on cards to determine whether to render viewer.
- Adds hidden showToAll setting for use with cardDealer.draw().
- Adds settings for display on pass to hand, pass to pile, and draw.
- Adds PseudoCard class for constructing card data structure for images.

kristianserrano committed:
Updates documentation
- Macros still need to be updated.

kristianserrano committed:
Merge branch 'master' into draw-support

# Conflicts:
#	dist/module/cardDealer.mjs
#	dist/module/hooks.mjs

orcnog committed:
moves card back images into their own folder

kristianserrano commented:
@orcnog there is only one place where the code makes use of a hidden `showToAll` setting, and that's in `cardDealer.draw()` since it triggers the `cards.deal()` which triggers the hook, which triggers the display. After `cards.draw()` is executed and completed, it immediately sets the setting back to the default value which is false. In all other cases, a `showToAll` boolean value is passed from function to function.

orcnog committed:
update View Image macro with some new syntax

orcnog committed:
CardDealer and FancyDisplay constructors now expect object. Added border color & width and cardback settings. Made borderWidth configurable in macros/api

orcnog committed:
update README with Settings section

orcnog committed:
version bump

also enhanced the versionbump.js node script itself to now update package.json and README.md on ver bump.

kristianserrano committed:
Removes remaining reference to faceDown

orcnog committed:
Update README.md

whitespace fix

orcnog committed:
Update module.json

min version should be 10, not 9

orcnog committed:
Merge pull request #23 from orcnog/develop

Develop to Master
Lots of updates, v10 support, new settings, a bug fix (Issue #18), configurable border widths and colors, some internal architecture updates.

orcnog commented:
Hate to make you do this, but as I just pushed up a whole bunch of changes, could you redirect this PR to a new branch I just cut for you? https://github.com/orcnog/orcnog-card-viewer/tree/draw-pass-support

The branch is updated with a few of your changes from the v10 branch, but not all.  There will likely be some merge conflict to resolve, but at this point, it's the latest code, so I want to make sure we're keeping our feature branches up-to-date.  Thanks!
@kristianserrano

kristianserrano commented:
No problem.

kristianserrano commented:
I'll need to take some time with this. There's some new features (e.g. default card backs) and functional changes that I want to make sure are still working after the merge. In fact, the default card backs makes some logic easier, but I think you already knew that. ;)

orcnog committed:
add a new demo gif showcasing the 2 main macros, and update README

also remove legacy script, unused.

orcnog commented:
> @orcnog there is only one place where the code makes use of a hidden showToAll setting, and that's in cardDealer.draw()

So, still striving for a way to do it without caching a temp value globally, I experimented with the error coming from setting a custom `action` string, and realized that if I simply pass `chatNotification: false`, it still allows me to pass a custom action string and read it in the hook, but it doesn't throw the error... at least not in v11.  **Would you mind testing this in v10, @kristianserrano?**

Example:
```
await deck.deal([pile], 1, {
   how: CONST.CARD_DRAW_MODES.RANDOM,
   action: shareToAll ? 'deal orcnog_card_viewer_doshare' : 'deal',
   chatNotification: false
});
```
Of course, it suppresses the actual chat post when I do that, but I'm willing to give up the out-of-the-box chat notifications for card draws (only those originating from macros) if it makes the mod fully compatible with v10.  Actually, if it does work error-free in v10, maybe we just wrap it in a version detection condition and only suppress chatNotification in v10 (since v11 seems to throw an error but push through and complete the action anyway).

kristianserrano commented:
That's not a bad idea. Do you have that code in a branch I can check out?

orcnog commented:
~Working on one... might not be up for a few hours though.  My REAL job is getting in the way =)~

Scratch that.  I just pushed that one change up to [develop](https://github.com/orcnog/orcnog-card-viewer/tree/develop), and merged it into [draw-pass-support](https://github.com/orcnog/orcnog-card-viewer/tree/draw-pass-support) for you.

orcnog committed:
attempt to fix v10 .deal() erroring out due to custom 'action' string

this suppresses chat notifications for macro-originating card draws, but if it makes v10 work, i'm ok with that.

orcnog committed:
Merge pull request #25 from orcnog/develop

suppress chatNotification in .draw()

kristianserrano commented:
No worries. I'm in the same boat. I still have to finish resolving the merge conflicts. I might cancel the merge for now though.

kristianserrano commented:
> ~Working on one... might not be up for a few hours though.  My REAL job is getting in the way =)~
> 
> Scratch that.  I just pushed that one change up to [develop](https://github.com/orcnog/orcnog-card-viewer/tree/develop), and merged it into [draw-pass-support](https://github.com/orcnog/orcnog-card-viewer/tree/draw-pass-support) for you.

You rock.

orcnog committed:
Adds localization to all remaining client-facing or language-functional strings

orcnog committed:
version bump. and implements one last localization string in context. Couple of README tweaks.

orcnog committed:
Merge pull request #27 from orcnog/develop

Develop

kristianserrano committed:
Adds settings for draw and pass to hand and pile

kristianserrano committed:
Merge branch 'draw-pass-support' into draw-support

kristianserrano committed:
Merge branch 'v10-support' into draw-support
